pyccel/parser/semantic.py

```diff
@@ -3455,53 +3455,84 @@ def _visit_For(self, expr):
             for_expr.end_annotation = expr.end_annotation
         return for_expr
 
-
     def _visit_FunctionalFor(self, expr):
-        old_index   = expr.index
-        new_index   = self.scope.get_new_name()
-        expr.substitute(old_index, new_index)
+        """
+        Visit and transform a FunctionalFor AST node into an equivalent code block.
 
-        target  = expr.expr
-        index   = new_index
-        indices = [self.scope.get_expected_name(i) for i in expr.indices]
-        dims    = []
-        body    = expr.loops[1]
+        This method processes a `FunctionalFor` expression and transforms the loop structure
+        into a corresponding code block.
 
-        idx_subs = {}
-        #scope = self.create_new_loop_scope()
+        Parameters
+        ----------
+        expr : pyccel.ast.functionalexpr.FunctionalFor
+            The FunctionalFor AST node.
 
-        # The symbols created to represent unknown valued objects are temporary
+        Returns
+        -------
+        pyccel.ast.basic.CodeBlock
+            A code block containing the equivalent loops and necessary variable allocations for the given `FunctionalFor` expression.
+        """
+         
+        target  = expr.expr
+        indices = []
+        dims = []
+        idx_subs = {}
         tmp_used_names = self.scope.all_used_symbols.copy()
+        body = expr.loops[0]
         i = 0
         while isinstance(body, For):
-
-            stop  = None
+            stop = None
             start = LiteralInteger(0)
-            step  = LiteralInteger(1)
-            var   = indices[i]
-            i += 1
-            a     = self._visit(body.iterable)
-            if isinstance(a, PythonRange):
+            step = LiteralInteger(1)
+            vars = []
+            a = Iterable(self._visit(body.iterable))
+            if isinstance(a.iterable, PythonRange):
```

Comment:
I don't remember the purpose of `var` here. For the rest is it not possible to use `a.to_range` to reduce the number of ifs?